### DISCUSSION TITLE
Why the installation process gets so complicated?

### DISCUSSION THREAD

**Discussion by tiahino** [pnp4nagios/pnp4nagios#55] [General]
Hi folks,

Not to start any flame war here, but I think the recently added configure/makefile changes to cover every possible scenario are not really necessary.

pnp4nagios is most likely is used in legacy installations, given that the libraries it is based upon (kohana) are long deceased. Most likely reasons for update are to get it compatible with recent php versions. Under this circumstances it is likely all the config files etc are already in place and the only thing necessary is the updated pnp4nagios code.

Seasoned Ops folks know, that keeping build environment on production hosts is a security concern, so most likely it should be packaged elsewhere and then the package deployed on the prod host in question. So assumption that all those apache directories and configurations are available during build will not hold. Especially if the package is compiled in some sort of a container only providing build s/w necessary to compile the stuff.

Also, if the build process (i.e. rpmbuild) is executed as a non-root user (as recommended) any attempt to change uid/gid on the objects being installed would fail. This is the task for a package management s/w (such as rpm or apt) - which is typically executed as root and thus is fully capable of changing the permissions according to the package specifications.

I'd also think that maintenance of these additions could be a challenging task too - given a wide variety of combinations used for deployment.

My suggestion is to reconsider deviating from the original packaging - which provides all those configuration options in the form of sample files provided along with the documentation which clearly describes possible deployment sceenarios. This should make the maintenance job easier and allow the developers to concentrate on the application (vs deployment) features

**Comment by celane** [comment]
In the first iteration of my work on pnp4nagios, I concentrated entirely in building rpms, because that's what I use. 

The more recent changes (last few months) I've tried to make "build from source" work properly, and that's resulted in a number of "install config" iterations, at least on my working source files; usually a result of getting an increased understanding of the capabilities and peculiarities of autoconf.  I test these on some virtual machines loaded with Alma, Debian, and NetBSD*, that let me test "fresh installs"; so yeah, it's easier for me to have the installation do all the config files.  (*NetBSD is still a problem)

But point taken, most people aren't doing "fresh installs"; I've now backed out most of that code that installs the config to a running system, at least on  my working copy of the code. 

Current task is getting the rpms building again (getting closer!) and then seeing what damage that did to the "build from source" config.  I expect that there's going to be some more 'testing branch' pre-releases before it's ready for a new 'main-branch' release.

**Reply by tiahino** [reply]
>  I expect that there's going to be some more 'testing branch' pre-releases before it's ready for a new 'main-branch' release.

In this case please integrate the fixes I provided in two issues I have opened - they address some broken functionality.

**Reply by celane** [reply]
fixes are in! 

Regarding "configuration" complexity: I *finally* got a configure.ac/Makefile.in that works for Alma, Debian, and NetBSD. That should cover most of the "unix on PC" world. But I backed off the "fullinstall messes with existing config"; most of that was so that I could quickly run through a build/install/test cycle on different virtual machines. 

So instead, the system-customized config files are in a pnp4nagios/sample_config/ directory, so you can use them or ignore them, as suits. 

rpm installs are rather nice, in that when updating config files you either have the old config moved to .rpmsave or the (proposed) new config in .rpmnew.  Perhaps there's a good way to get that functionality with Makefile installs (likely to be awkwardly non-standard) and other types of packaging systems (would have to learn them).  So for now, not touching other packages config files.

**Comment by celane** [comment]
closing this issue because installation simplified again.
